spark/src/main/java/org/apache/iceberg/spark/actions/BaseRewriteDataFilesSparkAction.java

```diff
@@ -0,0 +1,372 @@
+      commitUtil.commitOrClean(Sets.newHashSet(rewrittenGroups));
+      rewrittenGroups.forEach(group -> results.put(group.info(), group.asResult()));
+    } catch (ValidationException | CommitFailedException e) {
+      String errorMessage = String.format(
+          "Cannot commit rewrite because of a ValidationException or CommitFailedException. This usually means that " +
+              "this rewrite has conflicted with another concurrent Iceberg operation. To reduce the likelihood of " +
+              "conflicts, set %s which will break up the rewrite into multiple smaller commits controlled by %s. " +
+              "Separate smaller rewrite commits can succeed independently while any commits that conflict with " +
+              "another Iceberg operation will be ignored. This mode will create additional snapshots in the table " +
+              "history, one for each commit.",
+          PARTIAL_PROGRESS_ENABLED, PARTIAL_PROGRESS_MAX_COMMITS);
+      throw new RuntimeException(errorMessage, e);
+    }
+
+    return new BaseRewriteDataFilesResult(Maps.newHashMap(results));
+  }
+
+  private Result doExecuteWithPartialProgress(RewriteExecutionContext ctx, Stream<RewriteFileGroup> groupStream) {
+    ExecutorService rewriteService = rewriteService();
+
+    // Start Commit Service
+    int groupsPerCommit = IntMath.divide(ctx.totalGroupCount(), maxCommits, RoundingMode.CEILING);
+    RewriteDataFilesCommitUtil.CommitService commitService = commitUtil().service(groupsPerCommit);
+    commitService.start();
+
+    // Start rewrite tasks
+    Tasks.foreach(groupStream)
+        .suppressFailureWhenFinished()
+        .executeWith(rewriteService)
+        .noRetry()
+        .onFailure((fileGroup, exception) -> LOG.error("Failure during rewrite group {}", fileGroup.info(), exception))
+        .run(fileGroup -> commitService.offer(rewriteFiles(ctx, fileGroup)));
+    rewriteService.shutdown();
+
+    // Stop Commit service
+    commitService.close();
+    List<RewriteFileGroup> commitResults = commitService.results();
+    if (commitResults.size() == 0) {
+      LOG.error("{} is true but no rewrite commits succeeded. Check the logs to determine why the individual " +
+          "commits failed. If this is persistent it may help to increase {} which will break the rewrite operation " +
+          "into smaller commits.", PARTIAL_PROGRESS_ENABLED, PARTIAL_PROGRESS_MAX_COMMITS);
+    }
+
+    return new BaseRewriteDataFilesResult(commitResults.stream()
+        .collect(Collectors.toMap(RewriteFileGroup::info, RewriteFileGroup::asResult)));
+  }
+
+  private Stream<RewriteFileGroup> toGroupStream(RewriteExecutionContext ctx,
+                                                 Map<StructLike, List<List<FileScanTask>>> fileGroupsByPartition) {
+
+    // Todo Add intelligence to the order in which we do rewrites instead of just using partition order
+    return fileGroupsByPartition.entrySet().stream()
+        .flatMap(
+            e -> {
```

Comment:
nit: move `e -> {` to the line above and then reformat the block below? Will reduce the indentation in the closure.